server/src/test/java/org/elasticsearch/transport/LeakTrackerTests.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.transport;
+
+import com.carrotsearch.randomizedtesting.annotations.Name;
+import com.carrotsearch.randomizedtesting.annotations.ParametersFactory;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.core.AbstractRefCounted;
+import org.elasticsearch.core.Assertions;
+import org.elasticsearch.core.RefCounted;
+import org.elasticsearch.core.Releasables;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.test.ReachabilityChecker;
+import org.junit.Before;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.stream.Stream;
+
+public class LeakTrackerTests extends ESTestCase {
+
+    private static final Logger logger = LogManager.getLogger();
+
+    private final TrackedObjectLifecycle trackedObjectLifecycle;
+    private ReachabilityChecker reachabilityChecker;
+
+    @Before
+    public void createReachabilityTracker() {
+        reachabilityChecker = new ReachabilityChecker();
+    }
+
+    @Before
+    public void onlyRunWhenAssertionsAreEnabled() {
+        assumeTrue("Many of these tests don't make sense when assertions are disabled", Assertions.ENABLED);
+    }
+
+    public LeakTrackerTests(@Name("trackingMethod") TrackedObjectLifecycle trackedObjectLifecycle) {
+        this.trackedObjectLifecycle = trackedObjectLifecycle;
+    }
+
+    @ParametersFactory(shuffle = false)
+    public static Iterable<Object[]> parameters() {
+        return Stream.of(
+            new PojoTrackedObjectLifecycle(),
+            new ReleasableTrackedObjectLifecycle(),
+            new ReferenceCountedTrackedObjectLifecycle()
+        ).map(i -> new Object[] { i }).toList();
+    }
+
+    public void testWillLogErrorWhenTrackedObjectIsNotClosed() throws Exception {
+        // Let it go out of scope without closing
+        trackedObjectLifecycle.createAndTrack(reachabilityChecker);
+        reachabilityChecker.ensureUnreachable();
+        assertBusy(() -> assertLeakDetected("LEAK: resource was not cleaned up before it was garbage-collected\\.(.*|\\s)*"));
```

Comment:
This regex will match any `LeakTracker`-originated leak (it won't match those originating from netty). It'll probably end up being repeated in all such tests. I wonder if it's worth implementing an overload that takes no regex and matches any leaks, if you're testing some kind of leak detection you probably just care that one was detected?

Reply:
The only way to be more specific, currently, is by matching the stack-trace. We'd need to allow the callers of `LeakTracker#track` to provide something to be included in the message if we wanted to enable more specific matching on the message alone.

Reply:
I think it's fine to just assert that a leak was detected without needing to be precise about exactly which one it was. We can always revisit later if we need something more discriminating, but I expect that any tests that deliberately generate leaks won't really be the best place to detect other inadvertent leaks too.